File: my-submissions/0345-reverse-vowels-of-a-string/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
    // set for faster lookup
    // let vowels = new Set(['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']);
    let vowels = 'aeiouAEIOU';
    const arr = s.split('');
    let left = 0, right = arr.length - 1

    while(left < right) {
        // for set
        // while(left < right && !vowels.has(arr[left])) left++;
        // while(left < right && !vowels.has(arr[right])) right--;

        while(left < right && !vowels.includes(arr[left])) left++;
        while(left < right && !vowels.includes(arr[right])) right--;
        [arr[left], arr[right]] = [arr[right], arr[left]];
        left++;
        right--;
    }
    // console.log(arr);
    return arr.join('');
};
// input : IceCreAm

File: my-submissions/0345-reverse-vowels-of-a-string/solution.java
class Solution {
    public String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<>();
        for (char c : "aeiouAEIOU".toCharArray()) {
            vowels.add(c);
        }

        char[] arr = s.toCharArray();
        int left = 0;
        int right = arr.length - 1;

        while (left < right) {
            // Move left pointer until a vowel is found
            while (left < right && !vowels.contains(arr[left])) {
                left++;
            }
            // Move right pointer until a vowel is found
            while (left < right && !vowels.contains(arr[right])) {
                right--;
            }

            // Swap the vowels
            char temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;

            left++;
            right--;
        }

        return new String(arr);
    }
}
